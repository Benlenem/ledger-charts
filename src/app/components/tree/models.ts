import { Observable } from "rxjs/Observable";

export interface TreeItem {
    id: string;
    title: string;
    subtitle: string;
    isChecked: boolean;
    childrenIds: string[];
}

export interface TreeDatasource {
    getItemForId(id: string): Observable<TreeItem | undefined>;
}

export interface TreeDelegate {
    onItemClicked(item: TreeItem);
    onItemChecked(item: TreeItem, isChecked: boolean);
}