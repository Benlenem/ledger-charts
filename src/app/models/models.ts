import * as moment from "moment"
import Decimal from "decimal.js"

export interface Transaction {
    uuid?: string
    header: Header
    postings: Array<Posting>
}

interface Header {
    date: moment.Moment
    title: string
    tag?: string
}

export interface Posting {
    tag?: string
    account: string
    amount?: decimal.Decimal
    currency?: string
    comment?: string
}

export class Account {
    name: string
    children: Set<Account>;
    balance: decimal.Decimal
    childrenBalance: decimal.Decimal
    nbTransactions: number
    nbChildrenTransactions: number
    debits: decimal.Decimal
    credits: decimal.Decimal
    childrenDebits: decimal.Decimal
    childrenCredits: decimal.Decimal

    constructor(n: string) {
        this.name = n;
        this.balance = new Decimal(0)
        this.childrenBalance = new Decimal(0)
        this.nbTransactions = 0
        this.nbChildrenTransactions = 0
        this.debits = new Decimal(0)
        this.credits = new Decimal(0)
        this.childrenDebits = new Decimal(0)
        this.childrenCredits = new Decimal(0)
        this.children = new Set()
    }
}