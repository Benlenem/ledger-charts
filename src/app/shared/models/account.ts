import Decimal from 'decimal.js';

export class Account {
    name: string;
    children: Set<Account>;
    balance: decimal.Decimal;
    childrenBalance: decimal.Decimal;
    nbTransactions: number;
    nbChildrenTransactions: number;
    debits: decimal.Decimal;
    credits: decimal.Decimal;
    childrenDebits: decimal.Decimal;
    childrenCredits: decimal.Decimal;

    constructor(n: string) {
        this.name = n;
        this.balance = new Decimal(0);
        this.childrenBalance = new Decimal(0);
        this.nbTransactions = 0;
        this.nbChildrenTransactions = 0;
        this.debits = new Decimal(0);
        this.credits = new Decimal(0);
        this.childrenDebits = new Decimal(0);
        this.childrenCredits = new Decimal(0);
        this.children = new Set();
    }
}
