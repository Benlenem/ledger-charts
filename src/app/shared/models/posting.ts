export interface Posting {
  tag?: string;
  account: string;
  amount?: string;
  currency?: string;
  comment?: string;
  readonly tags: string[],
}
