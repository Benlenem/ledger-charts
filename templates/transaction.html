<div class="cell">
    <span (click)="editDate = true" *ngIf="!editDate">{{transaction.header.date}}</span>
    <input type="text" (click)="editDate = false" *ngIf="editDate" value="{{transaction.header.date}}"/>
</div>

<div class="cell">
    <span (click)="editTag = true" *ngIf="!editTag">{{transaction.header.tag || 'Vide'}}</span>
    <select (click)="editTag = false" *ngIf="editTag">
        <option>Vide</option>
        <option>*</option>
        <option>!</option>
    </select>
</div>

<div class="cell">
    <span (click)="editTitle = true" *ngIf="!editTitle">{{transaction.header.title}}</span>
    <input type="text" (click)="editTitle = false" *ngIf="editTitle" value="{{transaction.header.title}}"/>
</div>

<div class="cell cell50">
    <div *ngFor="let p of transaction.postings" class="row">
        <posting [posting]=p></posting>
    </div>
</div>