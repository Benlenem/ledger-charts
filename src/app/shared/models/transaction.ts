import { Posting } from './posting';

export interface Transaction {
  readonly header: Header;
  readonly postings: Array<Posting>;
}

interface Header {
  readonly date: number;
  readonly title: string;
  readonly tag?: string;
}

export interface TransactionWithUUID extends Transaction {
  readonly uuid: string;
}
