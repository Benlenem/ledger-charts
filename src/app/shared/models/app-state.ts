import { Transaction, TransactionWithUUID } from './transaction';
import { ReduxAccount } from './account';
import * as moment from 'moment';

export type TransactionMap = { [k: string]: TransactionWithUUID };
export type AccountMap = { [k: string]: ReduxAccount };

export interface AppState {
  entities: {
    accounts: AccountMap;
    transactions: TransactionMap;
  };
  ui: {
    selectedAccounts: string[];
    rootAccount: string;
    editedTransaction?: string;
    isLeftMenuOpen: boolean;
    isTransactionPanelOpen: boolean;
    persistedAt?: moment.Moment;
    isLoading: boolean,
  };
}
