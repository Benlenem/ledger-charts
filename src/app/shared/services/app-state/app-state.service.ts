import { Injectable } from '@angular/core';
import Decimal from 'decimal.js';
import * as moment from 'moment';
import 'rxjs/add/observable/merge';
import 'rxjs/add/operator/merge';
import { BehaviorSubject } from 'rxjs/BehaviorSubject';

import { Observable } from 'rxjs/Observable';
import { Subject } from 'rxjs/Subject';
import {v4 as uuid } from 'uuid';
import { Account } from '../../models/account';
import { Transaction, TransactionWithUUID } from '../../models/transaction';
import 'rxjs/add/observable/of';
import 'rxjs/add/observable/concat';
import 'rxjs/add/observable/from';

const ROOT_ACCOUNT = 'ROOT';

@Injectable()
export class AppStateService {

  private _selectedAccounts: Set<Account> = new Set();
  private _rootAccount: Account = new Account(ROOT_ACCOUNT);
  private _editedTransaction?: TransactionWithUUID;
  private _transactions: Map<string, TransactionWithUUID> = new Map();
  private _selectedAccountsSubject: Subject<Set<Account>> = new BehaviorSubject (this._selectedAccounts);
  private _rootAccountSubject: Subject<Account | undefined> = new BehaviorSubject(undefined);
  private _editedTransactionSubject: Subject<TransactionWithUUID | undefined> = new BehaviorSubject(undefined);
  private _transactionsChangedSubject: Subject<TransactionWithUUID[]> = new Subject();

  /****** HOT OBSERVABLES ******/

  selectedAccountsHotObservable(): Observable<Set<Account>> {
     return this._selectedAccountsSubject.asObservable().do( a => console.log("SIZE " + a.size));
  }

  selectedTransactionsHotObservable(): Observable<TransactionWithUUID[]> {
    const modif = this._transactionsChangedSubject.flatMap( () => {
      console.log('EUUUUUH');
      return this._selectedAccountsSubject;
    } );
    const selectedAccounts = this._selectedAccountsSubject.do(t => console.log('uh'));

    return Observable.merge(selectedAccounts, modif)
    .map( accounts => {
      console.log(accounts);
        const accountsNames = Array.from(accounts).map(a => a.name);
        const t =  this.getTransactionsUsingAccounts(accountsNames);
        console.log(t);
        return t;
      });
  }

  rootAccountHotObservable(): Observable<Account | undefined> {
    return this._rootAccountSubject.asObservable();
  }

  editedTransactionHotObservable(): Observable<TransactionWithUUID | undefined> {
    return this._editedTransactionSubject.asObservable();
  }

  allAccountsFlattenedHotObservable(): Observable<Account[]> {
    return this._rootAccountSubject.flatMap( root => Observable.of(root ? [root, ...this.allChildAccounts(root)] : []));
  }

  transactionsChangedHotObservable(): Observable<TransactionWithUUID[]> {
    return this._transactionsChangedSubject.asObservable();
  }

  /****** COLD OBSERVABLES ******/

  setEditedTransactionColdObservable(transaction?: TransactionWithUUID): Observable<any> {
    return Observable.create( obs => {
      this._editedTransaction = transaction;
      this._editedTransactionSubject.next(this._editedTransaction);
      obs.complete();
    });
  }

  private selectAccountsColdObservable(isSelected: boolean, accounts: Account[], clearPrevious): Observable<any> {
    return Observable.create( obs => {
      console.log('SELECTTTTTTT');

      if (clearPrevious) {
        this._selectedAccounts.clear();
      }

      const children = accounts.map(a => this.allChildAccounts(a)).reduce( (a, b) => [...a, ...b], []);
      const accountsToSelect: Account[] = [...accounts, ...children];

      if (isSelected) {
        accountsToSelect.forEach( a => this._selectedAccounts.add(a) );
      } else {
        accountsToSelect.forEach( a => this._selectedAccounts.delete(a) );
      }

      const accountsCount = this.allChildAccounts(this._rootAccount).length;
      if (this._selectedAccounts.size === accountsCount && !this._selectedAccounts.has(this._rootAccount)) {
        this._selectedAccounts.add((this._rootAccount));
      } else if (this._selectedAccounts.size === 1 && this._selectedAccounts.has(this._rootAccount)) {
        this._selectedAccounts.delete(this._rootAccount);
      }

      this._selectedAccountsSubject.next(this._selectedAccounts);

      obs.complete();
    });
  }

  selectAccountColdObservable(isSelected: boolean, account: Account): Observable<any> {
    return this.selectAccountsColdObservable(isSelected, [account], false);
  }

  allTransactionsColdObservable(): Observable<TransactionWithUUID[]> {
    return Observable.of(Array.from(this._transactions.values()));
  }

  addOrUpdateTransactionsColdObservable(transactions: TransactionWithUUID[], append: boolean = true): Observable<any> {
    const o = Observable.create( obs => {
      if (!append) {
        this._transactions.clear();
      }
      transactions.forEach(tr => {
        this._transactions.set(tr.uuid, tr);
      });

      this._transactionsChangedSubject.next(transactions);

      obs.complete();
    });

    return Observable.concat(o, this.generateAccounts(), this.refreshSelectedAccounts());
  }

  deleteTransactionColdObservable(transaction: TransactionWithUUID): Observable<any> {
    const o = Observable.create( obs => {
      this._transactions.delete(transaction.uuid);
      this._transactionsChangedSubject.next([transaction]);

      if (this._editedTransaction === transaction) {
        this._editedTransactionSubject.next(undefined);
      }
      obs.complete();
    });
    return Observable.concat(o, this.generateAccounts(), this.refreshSelectedAccounts());
  }

  addTransactionsColdObservable(transactions: Transaction[], append: boolean = true): Observable<any> {
    return Observable.from(transactions)
      .map(this.addUUIDToTransaction)
      .toArray()
      .flatMap(tr => this.addOrUpdateTransactionsColdObservable(tr, append));
  }

  /****** Private ******/

  private getTransactionsUsingAccounts(accountsNames: string[]) {
    const allTransactions = Array.from(this._transactions.values());

    if (accountsNames.find(n => n === ROOT_ACCOUNT)) {
      return allTransactions;
    }

    return allTransactions
      .filter(tr => {
      return tr.postings.some( p => accountsNames.indexOf(p.account) !== -1 );
    });
  }

  private allChildAccounts(root: Account): Account[] {
    const children: Account[] = [];
    root.children.forEach(c => children.push(...[c, ...this.allChildAccounts(c)]));
    return children;
  }

  private generateAccounts(): Observable<any> {
    return this.createAccountsFromTransactions()
      .do( accounts => {
        const root = new Account(ROOT_ACCOUNT);
        root.children = new Set(accounts);
        this._rootAccount = root;
        this._rootAccountSubject.next(this._rootAccount);
      });
  }

  private refreshSelectedAccounts(): Observable<any> {
    return Observable.create( obs => {
      if (this._selectedAccounts.size > 0) {

        const previousSelectedAccounts = Array.from(this._selectedAccounts.values()).map(a => a.name);
        const newlySelectedAccounts = [this._rootAccount, ...this.allChildAccounts(this._rootAccount)]
          .filter(a => previousSelectedAccounts.includes(a.name));

        obs.next(newlySelectedAccounts);
      }
      obs.complete();
    }).flatMap(accounts => this.selectAccountsColdObservable(true, accounts, true));
  }

  private addMissingAmount(transaction: Transaction) {
    const t = transaction.postings.find(p => p.amount === undefined);

    if (!t) {
      return;
    }

    const sum = transaction.postings
      .filter(p => p.amount !== undefined)
      .map(p => p.amount)
      .reduce((a1, a2) => Decimal.add(a1!, a2!), new Decimal(0));

    t.amount = new Decimal(-sum!);
  }

  private createAccountsFromTransactions(): Observable<Account[]> {
    return Observable.create( obs => {
      const flatAccounts = new Map();

      Array.from(this._transactions.values()).forEach(tr => {
        this.addMissingAmount(tr);
          tr.postings.forEach(ps => {
              const accountParts = ps.account.split(':');
              let lastParent: Account | undefined;
              let currentAccountName = '';
              for (const part of accountParts) {
                  currentAccountName += part;
                  const account = this.getOrCreateAccount(currentAccountName, flatAccounts);
                  this.addAmountToAccount(account, ps.amount || new Decimal(0), currentAccountName === ps.account);

                  if (lastParent) {
                      lastParent.children.add(account);
                  }

                  lastParent = account;
                  currentAccountName += ':';
              }
          });
      });

      obs.next(this.topAccounts(flatAccounts));
      obs.complete();
    });
  }

  private getOrCreateAccount(name: string, accountsMap: Map<string, Account>): Account {
    let stat = accountsMap.get(name);

    if (!stat) {
        stat = new Account(name);
        accountsMap.set(name, stat);
    }

    return stat;
  }

  private addAmountToAccount(a: Account, amount: Decimal, isFinalAccount: boolean) {
    if (!isFinalAccount) {
        a.childrenBalance = a.childrenBalance.plus(amount);
        a.nbChildrenTransactions ++;
        if (amount.greaterThan(0)) {
            a.childrenCredits = a.childrenCredits.plus(amount);
        } else {
            a.childrenDebits = a.childrenDebits.plus(amount);
        }
    } else {
        a.balance = a.balance.plus(amount);
        a.nbTransactions++;

        if (amount.greaterThan(0)) {
          a.credits = a.credits.plus(amount);
        } else {
            a.debits = a.debits.plus(amount);
        }
    }
  }

  private topAccounts(accounts: Map<string, Account>): Account[] {
    return Array.from(accounts.values()).filter( a => a.name.indexOf(':') === -1).sort( (a1, a2) => a1.name.localeCompare(a2.name) );
  }

  private addUUIDToTransaction(transaction: Transaction): TransactionWithUUID {
    return {
      ...transaction,
      uuid: uuid(),
    };
  }
}
