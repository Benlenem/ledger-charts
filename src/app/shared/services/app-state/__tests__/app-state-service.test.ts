import { AppStateService } from '../app-state.service';
import 'rxjs/add/operator/toPromise';
import 'rxjs/add/observable/zip';
import 'rxjs/add/operator/mergeMap';
import 'rxjs/add/operator/do';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/first';
import { transactions, transactionWithNestedAccounts } from './fixtures';
import 'rxjs/add/operator/toArray';
import 'rxjs/add/operator/take';
import { Transaction, TransactionWithUUID } from '../../../models/transaction';
import Decimal from 'decimal.js';
import * as moment from 'moment';

jest.mock('uuid', () => {
  let counter = 0;
  const uuids = Array.from(Array(20).keys()).map(u => `uuid ${u}`);

  return {
    v4: jest.fn(() => {
      const uuid = uuids[counter % (uuids.length - 1)];
      counter++;
      return uuid;
    }),
  };
});

interface State {
  allAccountsFlattened: Account[];
  rootAccount: Account;
  allTransactions: TransactionWithUUID[];
  selectedAccounts: Set<Account>;
  editedTransaction: Transaction | undefined;
  selectedTransactions: TransactionWithUUID[];
  transactionsChanged: TransactionWithUUID[];
}

const getState = async (service: AppStateService) => {
  const allAccountsFlattened = await service
    .allAccountsFlattenedHotObservable()
    .take<Account[]>(1)
    .toPromise();
  const rootAccount = await service
    .rootAccountHotObservable()
    .take<Account | undefined>(1)
    .toPromise();
  const allTransactions = await service.allTransactionsColdObservable().toPromise();
  const selectedAccounts = await service
    .selectedAccountsHotObservable()
    .take<Set<Account>>(1)
    .toPromise();
  const editedTransaction = await service
    .editedTransactionHotObservable()
    .take<TransactionWithUUID | undefined>(1)
    .toPromise();
  const selectedTransactions = await service
    .selectedTransactionsHotObservable()
    .take<TransactionWithUUID[]>(1)
    .toPromise();
  const transactionsChanged = await service
    .transactionsChangedHotObservable()
    .take<TransactionWithUUID[]>(1)
    .toPromise();

  return {
    allAccountsFlattened,
    rootAccount,
    allTransactions,
    selectedAccounts,
    editedTransaction,
    selectedTransactions,
    transactionsChanged,
  };
};

describe(AppStateService.name, () => {
  it('has empty state at startup', async () => {
    const service = new AppStateService();
    const state = await getState(service);

    expect(state).toMatchSnapshot();
  });

  it('changes state when adding first transactions', async () => {
    // given
    const service = new AppStateService();

    // when
    await service.addTransactionsColdObservable(transactions, true).toPromise();

    // then
    const state = await getState(service);
    expect(state).toMatchSnapshot();
  });

  it('changes state when appending transactions', async () => {
    // given
    const service = new AppStateService();
    await service.addTransactionsColdObservable([transactions[0]], true).toPromise();

    // when
    await service.addTransactionsColdObservable([transactions[1]], false).toPromise();

    // then
    const state = await getState(service);
    expect(state).toMatchSnapshot();
  });

  it('changes state when adding nested accounts', async () => {
    // given
    const service = new AppStateService();

    // when
    await service.addTransactionsColdObservable([transactionWithNestedAccounts], true).toPromise();

    // then
    const state = await getState(service);
    expect(state).toMatchSnapshot();
  });

  it('changes state when replacing one transaction with another', async () => {
    // given
    const service = new AppStateService();
    await service.addTransactionsColdObservable([transactions[0]], true).toPromise();

    // when
    await service.addTransactionsColdObservable([transactions[1]], true).toPromise();

    // then
    const state = await getState(service);
    expect(state).toMatchSnapshot();
  });

  it('changes state when replacing one transaction with nothing', async () => {
    // given
    const service = new AppStateService();
    await service.addTransactionsColdObservable([transactions[0]], true).toPromise();

    // when
    await service.addTransactionsColdObservable([], true).toPromise();

    // then
    const state = await getState(service);
    expect(state).toMatchSnapshot();
  });

  it('changes state when removing transactions', async () => {
    // given
    const service = new AppStateService();
    await service.addTransactionsColdObservable(transactions, true).toPromise();

    // when
    const allTransactions = await service.allTransactionsColdObservable().toPromise();
    await service.deleteTransactionColdObservable(allTransactions[0]).toPromise();

    // then
    const state = await getState(service);
    expect(state).toMatchSnapshot();
  });

  it('changes state when editing a transaction', async () => {
    // given
    const service = new AppStateService();
    await service.addTransactionsColdObservable(transactions, true).toPromise();

    // when
    const allTransactions = await service.allTransactionsColdObservable().toPromise();
    await service.setEditedTransactionColdObservable(allTransactions[0]).toPromise();

    // then
    const state = await getState(service);
    expect(state).toMatchSnapshot();
  });

  it('changes state when updating a transaction', async () => {
    // given
    const service = new AppStateService();
    await service.addTransactionsColdObservable(transactions, true).toPromise();

    // when
    const allTransactions = await service.allTransactionsColdObservable().toPromise();
    const modifiedTransaction = {
      ...allTransactions[0],
      header: {
        title: 'New title',
        date: moment.utc('20111031', 'YYYYMMDD'),
      },
      postings: [
        {
          account: 'Account C',
          amount: new Decimal(30),
        },
        {
          account: 'Account D',
        },
      ],
    };
    await service.updateTransactionColdObservable(modifiedTransaction).toPromise();

    // then
    const state = await getState(service);
    expect(state).toMatchSnapshot();
  });

  it('changes state when selecting root account', async () => {
    // given
    const service = new AppStateService();
    await service.addTransactionsColdObservable(transactions, true).toPromise();

    // when
    const rootAccount = await service
      .rootAccountHotObservable()
      .take(1)
      .toPromise();
    await service.selectAccountColdObservable(true, rootAccount!).toPromise();

    // then
    const state = await getState(service);
    expect(state).toMatchSnapshot();
  });
});
