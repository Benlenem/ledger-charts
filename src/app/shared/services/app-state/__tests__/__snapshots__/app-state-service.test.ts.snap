// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppStateService addTransactionsColdObservable appends transactions 1`] = `
Array [
  Object {
    "header": Object {
      "date": "2013-02-08T00:00:00.000Z",
      "title": "Bakery",
    },
    "postings": Array [
      Object {
        "account": "Bank:Current account",
        "amount": "-3.05",
        "comment": "Comment",
        "currency": "$",
        "tag": "!",
      },
      Object {
        "account": "Expenses:Grocery:Bread",
        "amount": "3.05",
      },
    ],
    "uuid": "uuid 6",
  },
  Object {
    "header": Object {
      "date": "2013-02-10T00:00:00.000Z",
      "title": "Salary",
    },
    "postings": Array [
      Object {
        "account": "Bank:Current account",
        "amount": "500",
        "currency": "€",
      },
      Object {
        "account": "Incomes:Salary",
        "amount": "-500",
      },
    ],
    "uuid": "uuid 7",
  },
]
`;

exports[`AppStateService addTransactionsColdObservable empties selected accounts 1`] = `Set {}`;

exports[`AppStateService addTransactionsColdObservable replaces transactions 1`] = `
Array [
  Object {
    "header": Object {
      "date": "2013-02-10T00:00:00.000Z",
      "title": "Salary",
    },
    "postings": Array [
      Object {
        "account": "Bank:Current account",
        "amount": "500",
        "currency": "€",
      },
      Object {
        "account": "Incomes:Salary",
        "amount": "-500",
      },
    ],
    "uuid": "uuid 5",
  },
]
`;

exports[`AppStateService addTransactionsColdObservable sets root account 1`] = `
Account {
  "balance": "0",
  "children": Set {
    Account {
      "balance": "0",
      "children": Set {
        Account {
          "balance": "496.95",
          "children": Set {},
          "childrenBalance": "0",
          "childrenCredits": "0",
          "childrenDebits": "0",
          "credits": "500",
          "debits": "-3.05",
          "name": "Bank:Current account",
          "nbChildrenTransactions": 0,
          "nbTransactions": 2,
        },
      },
      "childrenBalance": "496.95",
      "childrenCredits": "500",
      "childrenDebits": "-3.05",
      "credits": "0",
      "debits": "0",
      "name": "Bank",
      "nbChildrenTransactions": 2,
      "nbTransactions": 0,
    },
    Account {
      "balance": "0",
      "children": Set {
        Account {
          "balance": "0",
          "children": Set {
            Account {
              "balance": "3.05",
              "children": Set {},
              "childrenBalance": "0",
              "childrenCredits": "0",
              "childrenDebits": "0",
              "credits": "3.05",
              "debits": "0",
              "name": "Expenses:Grocery:Bread",
              "nbChildrenTransactions": 0,
              "nbTransactions": 1,
            },
          },
          "childrenBalance": "3.05",
          "childrenCredits": "3.05",
          "childrenDebits": "0",
          "credits": "0",
          "debits": "0",
          "name": "Expenses:Grocery",
          "nbChildrenTransactions": 1,
          "nbTransactions": 0,
        },
      },
      "childrenBalance": "3.05",
      "childrenCredits": "3.05",
      "childrenDebits": "0",
      "credits": "0",
      "debits": "0",
      "name": "Expenses",
      "nbChildrenTransactions": 1,
      "nbTransactions": 0,
    },
    Account {
      "balance": "0",
      "children": Set {
        Account {
          "balance": "-500",
          "children": Set {},
          "childrenBalance": "0",
          "childrenCredits": "0",
          "childrenDebits": "0",
          "credits": "0",
          "debits": "-500",
          "name": "Incomes:Salary",
          "nbChildrenTransactions": 0,
          "nbTransactions": 1,
        },
      },
      "childrenBalance": "-500",
      "childrenCredits": "0",
      "childrenDebits": "-500",
      "credits": "0",
      "debits": "0",
      "name": "Incomes",
      "nbChildrenTransactions": 1,
      "nbTransactions": 0,
    },
  },
  "childrenBalance": "0",
  "childrenCredits": "0",
  "childrenDebits": "0",
  "credits": "0",
  "debits": "0",
  "name": "ROOT",
  "nbChildrenTransactions": 0,
  "nbTransactions": 0,
}
`;

exports[`AppStateService allAccountsFlattenedHotObservable returns nothing at startup 1`] = `Array []`;

exports[`AppStateService allTransactionsColdObservable returns nothing at startup 1`] = `Array []`;

exports[`AppStateService allTransactionsColdObservable returns transactions 1`] = `
Array [
  Object {
    "header": Object {
      "date": "2013-02-08T00:00:00.000Z",
      "title": "Bakery",
    },
    "postings": Array [
      Object {
        "account": "Bank:Current account",
        "amount": "-3.05",
        "comment": "Comment",
        "currency": "$",
        "tag": "!",
      },
      Object {
        "account": "Expenses:Grocery:Bread",
        "amount": "3.05",
      },
    ],
    "uuid": "uuid 0",
  },
  Object {
    "header": Object {
      "date": "2013-02-10T00:00:00.000Z",
      "title": "Salary",
    },
    "postings": Array [
      Object {
        "account": "Bank:Current account",
        "amount": "500",
        "currency": "€",
      },
      Object {
        "account": "Incomes:Salary",
        "amount": "-500",
      },
    ],
    "uuid": "uuid 1",
  },
]
`;

exports[`AppStateService editedTransactionHotObservable returns edited transaction 1`] = `
Object {
  "header": Object {
    "date": "2013-02-08T00:00:00.000Z",
    "title": "Bakery",
  },
  "postings": Array [
    Object {
      "account": "Bank:Current account",
      "amount": "-3.05",
      "comment": "Comment",
      "currency": "$",
      "tag": "!",
    },
    Object {
      "account": "Expenses:Grocery:Bread",
      "amount": "3.05",
    },
  ],
  "uuid": "uuid 12",
}
`;

exports[`AppStateService rootAccountHotObservable returns account 1`] = `
Account {
  "balance": "0",
  "children": Set {
    Account {
      "balance": "0",
      "children": Set {
        Account {
          "balance": "496.95",
          "children": Set {},
          "childrenBalance": "0",
          "childrenCredits": "0",
          "childrenDebits": "0",
          "credits": "500",
          "debits": "-3.05",
          "name": "Bank:Current account",
          "nbChildrenTransactions": 0,
          "nbTransactions": 2,
        },
      },
      "childrenBalance": "496.95",
      "childrenCredits": "500",
      "childrenDebits": "-3.05",
      "credits": "0",
      "debits": "0",
      "name": "Bank",
      "nbChildrenTransactions": 2,
      "nbTransactions": 0,
    },
    Account {
      "balance": "0",
      "children": Set {
        Account {
          "balance": "0",
          "children": Set {
            Account {
              "balance": "3.05",
              "children": Set {},
              "childrenBalance": "0",
              "childrenCredits": "0",
              "childrenDebits": "0",
              "credits": "3.05",
              "debits": "0",
              "name": "Expenses:Grocery:Bread",
              "nbChildrenTransactions": 0,
              "nbTransactions": 1,
            },
          },
          "childrenBalance": "3.05",
          "childrenCredits": "3.05",
          "childrenDebits": "0",
          "credits": "0",
          "debits": "0",
          "name": "Expenses:Grocery",
          "nbChildrenTransactions": 1,
          "nbTransactions": 0,
        },
      },
      "childrenBalance": "3.05",
      "childrenCredits": "3.05",
      "childrenDebits": "0",
      "credits": "0",
      "debits": "0",
      "name": "Expenses",
      "nbChildrenTransactions": 1,
      "nbTransactions": 0,
    },
    Account {
      "balance": "0",
      "children": Set {
        Account {
          "balance": "-500",
          "children": Set {},
          "childrenBalance": "0",
          "childrenCredits": "0",
          "childrenDebits": "0",
          "credits": "0",
          "debits": "-500",
          "name": "Incomes:Salary",
          "nbChildrenTransactions": 0,
          "nbTransactions": 1,
        },
      },
      "childrenBalance": "-500",
      "childrenCredits": "0",
      "childrenDebits": "-500",
      "credits": "0",
      "debits": "0",
      "name": "Incomes",
      "nbChildrenTransactions": 1,
      "nbTransactions": 0,
    },
  },
  "childrenBalance": "0",
  "childrenCredits": "0",
  "childrenDebits": "0",
  "credits": "0",
  "debits": "0",
  "name": "ROOT",
  "nbChildrenTransactions": 0,
  "nbTransactions": 0,
}
`;

exports[`AppStateService rootAccountHotObservable returns nothing at startup 1`] = `undefined`;

exports[`AppStateService selectAccountsColdObservable selects accounts 1`] = `
Set {
  Account {
    "balance": "0",
    "children": Set {},
    "childrenBalance": "0",
    "childrenCredits": "0",
    "childrenDebits": "0",
    "credits": "0",
    "debits": "0",
    "name": "Account A",
    "nbChildrenTransactions": 0,
    "nbTransactions": 0,
  },
  Account {
    "balance": "0",
    "children": Set {},
    "childrenBalance": "0",
    "childrenCredits": "0",
    "childrenDebits": "0",
    "credits": "0",
    "debits": "0",
    "name": "Account B",
    "nbChildrenTransactions": 0,
    "nbTransactions": 0,
  },
}
`;

exports[`AppStateService selectAccountsColdObservable unselects accounts 1`] = `
Set {
  Account {
    "balance": "0",
    "children": Set {},
    "childrenBalance": "0",
    "childrenCredits": "0",
    "childrenDebits": "0",
    "credits": "0",
    "debits": "0",
    "name": "Account A",
    "nbChildrenTransactions": 0,
    "nbTransactions": 0,
  },
}
`;

exports[`AppStateService selectedTransactionsHotObservable automatically selects new transaction from selected account 1`] = `
Array [
  Object {
    "header": Object {
      "date": "2013-02-08T00:00:00.000Z",
      "title": "Bakery",
    },
    "postings": Array [
      Object {
        "account": "Bank:Current account",
        "amount": "-3.05",
        "comment": "Comment",
        "currency": "$",
        "tag": "!",
      },
      Object {
        "account": "Expenses:Grocery:Bread",
        "amount": "3.05",
      },
    ],
    "uuid": "uuid 18",
  },
  Object {
    "header": Object {
      "date": "2013-02-10T00:00:00.000Z",
      "title": "Salary",
    },
    "postings": Array [
      Object {
        "account": "Bank:Current account",
        "amount": "500",
        "currency": "€",
      },
      Object {
        "account": "Incomes:Salary",
        "amount": "-500",
      },
    ],
    "uuid": "uuid 0",
  },
]
`;

exports[`AppStateService selectedTransactionsHotObservable does not select new transaction from unselected account 1`] = `
Array [
  Object {
    "header": Object {
      "date": "2013-02-08T00:00:00.000Z",
      "title": "Bakery",
    },
    "postings": Array [
      Object {
        "account": "Bank:Current account",
        "amount": "-3.05",
        "comment": "Comment",
        "currency": "$",
        "tag": "!",
      },
      Object {
        "account": "Expenses:Grocery:Bread",
        "amount": "3.05",
      },
    ],
    "uuid": "uuid 1",
  },
]
`;

exports[`AppStateService selectedTransactionsHotObservable returns all transactions when root account is selected 1`] = `
Array [
  Object {
    "header": Object {
      "date": "2013-02-08T00:00:00.000Z",
      "title": "Bakery",
    },
    "postings": Array [
      Object {
        "account": "Bank:Current account",
        "amount": "-3.05",
        "comment": "Comment",
        "currency": "$",
        "tag": "!",
      },
      Object {
        "account": "Expenses:Grocery:Bread",
        "amount": "3.05",
      },
    ],
    "uuid": "uuid 14",
  },
  Object {
    "header": Object {
      "date": "2013-02-10T00:00:00.000Z",
      "title": "Salary",
    },
    "postings": Array [
      Object {
        "account": "Bank:Current account",
        "amount": "500",
        "currency": "€",
      },
      Object {
        "account": "Incomes:Salary",
        "amount": "-500",
      },
    ],
    "uuid": "uuid 15",
  },
]
`;

exports[`AppStateService selectedTransactionsHotObservable returns nothing at startup 1`] = `Array []`;

exports[`AppStateService selectedTransactionsHotObservable returns transaction when account is selected 1`] = `
Array [
  Object {
    "header": Object {
      "date": "2013-02-10T00:00:00.000Z",
      "title": "Salary",
    },
    "postings": Array [
      Object {
        "account": "Bank:Current account",
        "amount": "500",
        "currency": "€",
      },
      Object {
        "account": "Incomes:Salary",
        "amount": "-500",
      },
    ],
    "uuid": "uuid 17",
  },
]
`;
